File: scripts/demo_topology_schema.py
"""Demo: validate every committed topology fixture and print pass/fail.

Used by `just demo-topology-schema`. Exits non-zero if any fixture fails the
validation the fixture name claims (valid/ fixtures must validate;
topology-invalid/ fixtures must fail).
"""

from __future__ import annotations

from pathlib import Path

import yaml
from jsonschema import ValidationError
from swarmkit_schema import validate

REPO_ROOT = Path(__file__).resolve().parent.parent
FIXTURES = REPO_ROOT / "packages" / "schema" / "tests" / "fixtures"


def _load(path: Path) -> object:
    return yaml.safe_load(path.read_text(encoding="utf-8"))


def _run_case(path: Path, should_pass: bool) -> bool:
    tag = "✓" if should_pass else "✗"
    try:
        validate("topology", _load(path))
        actually_passed = True
    except ValidationError as exc:
        actually_passed = False
        reason = str(exc).splitlines()[0]

    ok = actually_passed == should_pass
    marker = "✓" if ok else "✗"
    if should_pass:
        note = "valid" if actually_passed else f"expected valid, got: {reason}"
    else:
        note = "rejected" if not actually_passed else "expected invalid, passed"
    print(f"  {marker} {tag} {path.name}  [{note}]")
    return ok


def main() -> int:
    valid_dir = FIXTURES / "topology"
    invalid_dir = FIXTURES / "topology-invalid"
    all_ok = True

    print("valid fixtures:")
    for p in sorted(valid_dir.glob("*.yaml")):
        all_ok &= _run_case(p, should_pass=True)

    print("invalid fixtures (should fail validation):")
    for p in sorted(invalid_dir.glob("*.yaml")):
        all_ok &= _run_case(p, should_pass=False)

    print()
    print("all cases passed." if all_ok else "one or more cases failed.")
    return 0 if all_ok else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/schema/typescript/scripts/demo-topology-schema.mjs
// Demo: validate every committed topology fixture from the TS validator.
// Used by `just demo-topology-schema`.
//
// Runs without requiring a prior build — imports ajv directly and reads the
// canonical JSON Schemas from their source-of-truth location. Equivalent to
// what `@swarmkit/schema`'s public API does.

import { readdirSync, readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import Ajv2020 from "ajv/dist/2020.js";
import addFormats from "ajv-formats";
import { parse as parseYaml } from "yaml";

const here = dirname(fileURLToPath(import.meta.url));
const SCHEMAS_ROOT = resolve(here, "..", "..", "schemas");
const FIXTURE_ROOT = resolve(here, "..", "..", "tests", "fixtures");

const topologySchema = JSON.parse(
  readFileSync(resolve(SCHEMAS_ROOT, "topology.schema.json"), "utf-8"),
);
const ajv = new Ajv2020({ strict: false, allErrors: true });
addFormats(ajv);
const validator = ajv.compile(topologySchema);

function load(kind, file) {
  return parseYaml(readFileSync(resolve(FIXTURE_ROOT, kind, file), "utf-8"));
}

function run(dir, shouldPass) {
  const files = readdirSync(resolve(FIXTURE_ROOT, dir))
    .filter((f) => f.endsWith(".yaml"))
    .sort();
  let allOk = true;
  for (const file of files) {
    const actuallyPassed = validator(load(dir, file)) === true;
    const ok = actuallyPassed === shouldPass;
    allOk = allOk && ok;
    const marker = ok ? "✓" : "✗";
    const tag = shouldPass ? "✓" : "✗";
    const note = shouldPass
      ? actuallyPassed
        ? "valid"
        : "expected valid, rejected"
      : actuallyPassed
        ? "expected invalid, passed"
        : "rejected";
    console.log(`  ${marker} ${tag} ${file}  [${note}]`);
  }
  return allOk;
}

console.log("valid fixtures:");
const validOk = run("topology", true);
console.log("invalid fixtures (should fail validation):");
const invalidOk = run("topology-invalid", false);

console.log();
const allOk = validOk && invalidOk;
console.log(allOk ? "all cases passed." : "one or more cases failed.");
process.exit(allOk ? 0 : 1);
